instrumentation/vertx-http-client/vertx-http-client-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v3_0/client/Vertx3HttpAttributesExtractor.java

```diff
@@ -5,16 +5,42 @@
 
 package io.opentelemetry.javaagent.instrumentation.vertx.v3_0.client;
 
+import io.opentelemetry.instrumentation.api.field.VirtualField;
 import io.opentelemetry.javaagent.instrumentation.vertx.client.AbstractVertxHttpAttributesExtractor;
 import io.vertx.core.http.HttpClientRequest;
 import javax.annotation.Nullable;
 
 final class Vertx3HttpAttributesExtractor extends AbstractVertxHttpAttributesExtractor {
+  private static final VirtualField<HttpClientRequest, VertxRequestInfo> requestInfoField =
+      VirtualField.find(HttpClientRequest.class, VertxRequestInfo.class);
 
   @Nullable
   @Override
   protected String url(HttpClientRequest request) {
-    return request.uri();
+    String uri = request.uri();
+    // Uri should be relative, but it is possible to misuse vert.x api and pass an absolute uri
+    // where relative is expected.
+    if (!isAbsolute(uri)) {
+      VertxRequestInfo requestInfo = requestInfoField.get(request);
+      uri = absoluteUri(requestInfo, uri);
+    }
+    return uri;
+  }
+
+  private static boolean isAbsolute(String uri) {
+    return uri.startsWith("http://") || uri.startsWith("https://");
+  }
+
+  private static String absoluteUri(VertxRequestInfo requestInfo, String uri) {
+    String result = requestInfo.isSsl() ? "https://" : "http://";
```

Comment:
This code is probably not called commonly since it is misuse of vertx API but we should probably still stick to StringBuilder as a general pattern for this sort of string building